Big Sur (macOS 11.0) compatibility

Latest version of the repo and tap fails to compile on Big Sur with the following error:

```
configure: error: Valgrind works on Darwin 10.x-19.x (Mac OS X 10.6-11 and macOS 10.12-15)
```

--- a/README.md
+++ b/README.md
@@ -6,17 +6,16 @@ This repository contains a version of Valgrind including a few patches to improv
 
 | Version                     | x86 | amd64 | arm64  | ppc    |
 | --------------------------- | --- | ----- | ------ | ------ |
-| macOS 10.13 and earlier[^1] | ✅  | ✅    | -      | ❌[^4] |
+| macOS 10.13 and earlier[^1] | ✅  | ✅    | -      | ❌[^3] |
 | macOS 10.14 (Mojave)        | ✅  | ✅    | -      | -      |
 | macOS 10.15 (Catalina)      | ✅  | ✅    | -      | -      |
-| macOS 11 (Big Sur)[^3]      | -   | ✅    | ❌[^2] | -      |
-| macOS 12 (Monterey)[^3]     | -   | ✅    | ❌[^2] | -      |
-| macOS 13 (Ventura)[^3]      | -   | ✅    | ❌[^2] | -      |
+| macOS 11 (Big Sur)          | -   | ✅    | ❌[^2] | -      |
+| macOS 12 (Monterey)         | -   | ✅    | ❌[^2] | -      |
+| macOS 13 (Ventura)          | -   | ✅    | ❌[^2] | -      |
 
 [^1]: Supported as part of upstream Valgrind.
 [^2]: Apple Silicon support in progress ([#56](https://github.com/LouisBrunner/valgrind-macos/issues/56))
-[^3]: macOS 11 and later doesn't track leak correctly ([#19](https://github.com/LouisBrunner/valgrind-macos/issues/19))
-[^4]: PowerPC is unsupported ([#62](https://github.com/LouisBrunner/valgrind-macos/issues/62))
+[^3]: PowerPC is unsupported ([#62](https://github.com/LouisBrunner/valgrind-macos/issues/62))
 
 Note that every version from macOS 10.12 onwards currently has the following issues:
 

--- a/coregrind/m_debuginfo/debuginfo.c
+++ b/coregrind/m_debuginfo/debuginfo.c
@@ -1832,6 +1832,60 @@ void VG_(di_notify_pdb_debuginfo)( Int fd_obj, Addr avma_obj,
 
 #endif /* defined(VGO_linux) || defined(VGO_darwin) || defined(VGO_solaris) || defined(VGO_freebsd) */
 
+#if defined(VGO_darwin) && DARWIN_VERS >= DARWIN_11_00
+// Special version of VG_(di_notify_mmap) specifically to read debug info from the DYLD Shared Cache (DSC)
+// We only use this on macOS 11.0 and later, because Apple stopped shipping dylib on-disk then.
+
+ULong VG_(di_notify_dsc)( const HChar* filename, Addr header, SizeT len )
+{
+   DebugInfo* di;
+
+   const Bool       debug = VG_(debugLog_getLevel)() >= 3;
+
+   if (debug)
+      VG_(dmsg)("di_notify_dsc-1: %s\n", filename);
+
+   if (!ML_(is_macho_object_file)( (const void*) header, len ))
+      return 0;
+
+   /* See if we have a DebugInfo for this filename.  If not,
+      create one. */
+   di = find_or_create_DebugInfo_for( filename );
+   vg_assert(di);
+
+   if (di->have_dinfo) {
+      if (debug)
+         VG_(dmsg)("di_notify_dsc-2x: "
+                   "ignoring mapping because we already read debuginfo "
+                   "for DebugInfo* %p\n", di);
+      return 0;
+   }
+
+   if (debug)
+      VG_(dmsg)("di_notify_dsc-2: "
+                "noting details in DebugInfo* at %p\n", di);
+
+   /* Note the details about the mapping. */
+   DebugInfoMapping map;
+   map.avma = header;
+   map.size = len;
+   map.foff = 0;
+   map.rx   = True;
+   map.rw   = False;
+   map.ro   = False;
+   VG_(addToXA)(di->fsm.maps, &map);
+
+   /* Update flags about what kind of mappings we've already seen. */
+   di->fsm.have_rx_map |= True;
+
+   vg_assert(!di->have_dinfo);
+
+   if (debug)
+      VG_(dmsg)("di_notify_dsc-3: "
+                "achieved accept state for %s\n", filename);
+   return di_notify_ACHIEVE_ACCEPT_STATE ( di );
+}
+#endif
 
 /*------------------------------------------------------------*/
 /*---                                                      ---*/

--- a/coregrind/m_debuginfo/debuginfo.c
+++ b/coregrind/m_debuginfo/debuginfo.c
@@ -683,7 +683,15 @@ static void check_CFSI_related_invariants ( const DebugInfo* di )
       been successfully read.  And that shouldn't happen until we have
       both a r-x and rw- mapping for the object.  Hence: */
    vg_assert(di->fsm.have_rx_map);
+#if defined(VGO_darwin) && DARWIN_VERS >= DARWIN_11_00
+   // not required, in the case of a DSC map we only have r-x
+   // (technically the DSC has multiple mappings but what's the point of adding all of them?)
+   if (di->fsm.have_ro_map) {
+     vg_assert(di->fsm.rw_map_count);
+   }
+#else
    vg_assert(di->fsm.rw_map_count);
+#endif
    for (i = 0; i < VG_(sizeXA)(di->fsm.maps); i++) {
       const DebugInfoMapping* map = VG_(indexXA)(di->fsm.maps, i);
       /* We are interested in r-x mappings only */
@@ -1839,7 +1847,6 @@ void VG_(di_notify_pdb_debuginfo)( Int fd_obj, Addr avma_obj,
 ULong VG_(di_notify_dsc)( const HChar* filename, Addr header, SizeT len )
 {
    DebugInfo* di;
-
    const Bool       debug = VG_(debugLog_getLevel)() >= 3;
 
    if (debug)

--- a/coregrind/m_debuginfo/image.c
+++ b/coregrind/m_debuginfo/image.c
@@ -583,8 +583,12 @@ static void set_CEnt ( const DiImage* img, UInt entNo, DiOffT off )
 
    if (img->source.is_local) {
       // Simple: just read it
+      if (img->source.fd == -1) {
+        VG_(memcpy)(&ce->data[0], ((const char *)img->source.session_id) + off, len);
+      } else {
       SysRes sr = VG_(pread)(img->source.fd, &ce->data[0], (Int)len, off);
       vg_assert(!sr_isError(sr));
+      }
    } else {
       // Not so simple: poke the server
       vg_assert(img->source.session_id > 0);
@@ -916,6 +920,39 @@ DiImage* ML_(img_from_fd)(Int fd, const HChar* fullpath)
    return img;
 }
 
+/* Create an image from a place in memory, this is to support certain use cases (DSC on macOS)
+   where images are already loaded in memory without changing every usage of DiImage. */
+DiImage* ML_(img_from_memory)(Addr a, SizeT size, const HChar* fullpath)
+{
+   if (size == 0 || size == DiOffT_INVALID
+       || /* size is unrepresentable as a SizeT */
+          size != (DiOffT)(SizeT)(size)) {
+      return NULL;
+   }
+
+   DiImage* img = ML_(dinfo_zalloc)("di.image.ML_iflf.1", sizeof(DiImage));
+   img->source.is_local   = True;
+   img->source.fd         = -1;
+   img->source.session_id = a; // FIXME: hacky, but avoids a new variable
+   img->size              = size;
+   img->real_size         = size;
+   img->ces_used          = 0;
+   img->source.name       = ML_(dinfo_strdup)("di.image.ML_iflf.2", fullpath);
+   img->cslc              = NULL;
+   img->cslc_size         = 0;
+   img->cslc_used         = 0;
+
+   /* Force the zeroth entry to be the first chunk of the file.
+      That's likely to be the first part that's requested anyway, and
+      loading it at this point forcing img->cent[0] to always be
+      non-empty, thereby saving us an is-it-empty check on the fast
+      path in get(). */
+   UInt entNo = alloc_CEnt(img, CACHE_ENTRY_SIZE, False/*!fromC*/);
+   vg_assert(entNo == 0);
+   set_CEnt(img, 0, 0);
+
+   return img;
+}
 
 
 /* Create an image from a file on a remote debuginfo server.  This is
@@ -1073,9 +1110,11 @@ void ML_(img_done)(DiImage* img)
 {
    vg_assert(img != NULL);
    if (img->source.is_local) {
+      if (img->source.fd != -1) {
       /* Close the file; nothing else to do. */
       vg_assert(img->source.session_id == 0);
       VG_(close)(img->source.fd);
+      }
    } else {
       /* Close the socket.  The server can detect this and will scrub
          the connection when it happens, so there's no need to tell it

--- a/coregrind/m_debuginfo/priv_image.h
+++ b/coregrind/m_debuginfo/priv_image.h
@@ -61,6 +61,8 @@ DiImage* ML_(img_from_local_file)(const HChar* fullpath);
 
 DiImage* ML_(img_from_fd)(Int fd, const HChar* fullpath);
 
+DiImage* ML_(img_from_memory)(Addr a, SizeT len, const HChar* fullpath);
+
 /* Create an image by connecting to a Valgrind debuginfo server
    (auxprogs/valgrind-di-server.c).  |filename| contains the object
    name to ask for; it must be a plain filename, not absolute, not a

--- a/coregrind/m_debuginfo/readmacho.c
+++ b/coregrind/m_debuginfo/readmacho.c
@@ -715,6 +715,7 @@ Bool ML_(read_macho_debug_info)( struct _DebugInfo* di )
    HChar*   dsymfilename = NULL;
    Bool     have_uuid    = False;
    Bool     from_memory  = False; // True if we're reading from DSC
+   Addr     kernel_slide = 0; // Used when from_memory is True
    UChar    uuid[16];
    Word     i;
    struct SEGMENT_COMMAND     link_edit = {.cmd = 0};
@@ -732,6 +733,7 @@ Bool ML_(read_macho_debug_info)( struct _DebugInfo* di )
    // (without adding tons of new functions and fields)
    if (di->fsm.rw_map_count == 0) {
      from_memory = True;
+     kernel_slide = VG_(dyld_cache_get_slide)();
    }
 #else
    vg_assert(di->fsm.rw_map_count);
@@ -970,7 +972,7 @@ Bool ML_(read_macho_debug_info)( struct _DebugInfo* di )
         // Then we get the offset of syms/strings within __LINKEDIT by removing the fileoffset
         // Then we add the __LINKEDIT address
         // Finally we calculate the proper offset of those addresses compared to the mach_header slice
-        Addr link_edit_addr = link_edit.vmaddr + VG_(dyld_cache_get_slide)();
+        Addr link_edit_addr = link_edit.vmaddr + kernel_slide;
         syms = ML_(cur_from_sli)(msli);
         syms.ioff = (link_edit_addr + (symcmd.symoff - link_edit.fileoff)) - rx_map->avma;
         strs = ML_(cur_from_sli)(msli);

--- a/coregrind/m_debuginfo/readmacho.c
+++ b/coregrind/m_debuginfo/readmacho.c
@@ -143,12 +143,19 @@ static void unmap_image ( /*MOD*/DiSlice* sli )
    returns DiSlice_INVALID if it fails.  If it succeeds, the returned
    slice is guaranteed to refer to a valid(ish) Mach-O image. */
 static DiSlice map_image_aboard ( DebugInfo* di, /* only for err msgs */
-                                  const HChar* filename )
+                                  const HChar* filename,
+                                  const DebugInfoMapping* rx_map )
 {
    DiSlice sli = DiSlice_INVALID;
 
    /* First off, try to map the thing in. */
-   DiImage* mimg = ML_(img_from_local_file)(filename);
+   DiImage* mimg;
+
+   if (rx_map != NULL) {
+     mimg = ML_(img_from_memory)(rx_map->avma, rx_map->size, filename);
+   } else {
+     mimg = ML_(img_from_local_file)(filename);
+   }
    if (mimg == NULL) {
       VG_(message)(Vg_UserMsg, "warning: connection to image %s failed\n",
                                filename );
@@ -703,6 +710,7 @@ Bool ML_(read_macho_debug_info)( struct _DebugInfo* di )
    DiCursor dysym_cur    = DiCursor_INVALID;
    HChar*   dsymfilename = NULL;
    Bool     have_uuid    = False;
+   Bool     from_memory  = False;
    UChar    uuid[16];
    Word     i;
    const DebugInfoMapping* rx_map = NULL;
@@ -714,7 +722,15 @@ Bool ML_(read_macho_debug_info)( struct _DebugInfo* di )
    /* This should be ensured by our caller (that we're in the accept
       state). */
    vg_assert(di->fsm.have_rx_map);
+#if DARWIN_VERS >= DARWIN_11_00
+   // FIXME: this is a hack to identify when a DebugInfo is associated with the DSC
+   // (without adding tons of new functions and fields)
+   if (di->fsm.rw_map_count == 0) {
+     from_memory = True;
+   }
+#else
    vg_assert(di->fsm.rw_map_count);
+#endif
 
    for (i = 0; i < VG_(sizeXA)(di->fsm.maps); i++) {
       const DebugInfoMapping* map = VG_(indexXA)(di->fsm.maps, i);
@@ -722,11 +738,11 @@ Bool ML_(read_macho_debug_info)( struct _DebugInfo* di )
          rx_map = map;
       if (map->rw && !rw_map)
          rw_map = map;
-      if (rx_map && rw_map)
+      if (rx_map && (rw_map || from_memory))
          break;
    }
    vg_assert(rx_map);
-   vg_assert(rw_map);
+   vg_assert(rw_map || from_memory);
 
    if (VG_(clo_verbosity) > 1)
       VG_(message)(Vg_DebugMsg,
@@ -735,7 +751,7 @@ Bool ML_(read_macho_debug_info)( struct _DebugInfo* di )
 
    VG_(memset)(&uuid, 0, sizeof(uuid));
 
-   msli = map_image_aboard( di, di->fsm.filename );
+   msli = map_image_aboard( di, di->fsm.filename, from_memory ? rx_map : NULL );
    if (!ML_(sli_is_valid)(msli)) {
       ML_(symerr)(di, False, "Connect to main image failed.");
       goto fail;
@@ -854,7 +870,7 @@ Bool ML_(read_macho_debug_info)( struct _DebugInfo* di )
                di->text_debug_bias = di->text_bias;
             }
             /* Try for __DATA */
-            if (!di->data_present
+            if (!from_memory && !di->data_present
                 && 0 == VG_(strcmp)(&seg.segname[0], "__DATA")
                 /* && DDD:seg->fileoff == 0 */ && seg.filesize != 0) {
                di->data_present = True;
@@ -1008,7 +1024,7 @@ Bool ML_(read_macho_debug_info)( struct _DebugInfo* di )
       if (VG_(clo_verbosity) > 1)
          VG_(message)(Vg_DebugMsg, "   dSYM= %s\n", dsymfilename);
 
-      dsli = map_image_aboard( di, dsymfilename );
+      dsli = map_image_aboard( di, dsymfilename, NULL );
       if (!ML_(sli_is_valid)(dsli)) {
          ML_(symerr)(di, False, "Connect to debuginfo image failed "
                                 "(first attempt).");
@@ -1078,7 +1094,7 @@ Bool ML_(read_macho_debug_info)( struct _DebugInfo* di )
       if (VG_(clo_verbosity) > 1)
          VG_(message)(Vg_DebugMsg, "   dsyms= %s\n", dsymfilename);
 
-      dsli = map_image_aboard( di, dsymfilename );
+      dsli = map_image_aboard( di, dsymfilename, NULL );
       if (!ML_(sli_is_valid)(dsli)) {
          ML_(symerr)(di, False, "Connect to debuginfo image failed "
                                 "(second attempt).");

--- a/coregrind/m_debuginfo/readmacho.c
+++ b/coregrind/m_debuginfo/readmacho.c
@@ -172,7 +172,7 @@ static DiSlice map_image_aboard ( DebugInfo* di, /* only for err msgs */
    DiOffT            fh_be_ioff = 0;
    struct fat_header fh_be;
    struct fat_header fh;
-     
+
    // Assume initially that we have a thin image, and narrow
    // the bounds if it turns out to be fat.  This stores |mimg| as
    // |sli.img|, so NULL out |mimg| after this point, for the sake of
@@ -306,7 +306,7 @@ static DiSlice map_image_aboard ( DebugInfo* di, /* only for err msgs */
    ML_(addSym) itself. */
 static
 void read_symtab( /*OUT*/XArray* /* DiSym */ syms,
-                  struct _DebugInfo* di, 
+                  struct _DebugInfo* di,
                   DiCursor symtab_cur, UInt symtab_count,
                   DiCursor strtab_cur, UInt strtab_sz )
 {
@@ -331,7 +331,7 @@ void read_symtab( /*OUT*/XArray* /* DiSym */ syms,
       } else {
          continue;
       }
-      
+
       if (di->trace_symtab) {
          HChar* str = ML_(cur_read_strdup)(
                          ML_(cur_plus)(strtab_cur, nl.n_un.n_strx),
@@ -383,7 +383,7 @@ void read_symtab( /*OUT*/XArray* /* DiSym */ syms,
       // and display as "(below main)".
       if (disym.pri_name[0] == '_') {
          disym.pri_name++;
-      } 
+      }
       else if (!VG_(clo_show_below_main) && VG_STREQ(disym.pri_name, "start")) {
          if (s_a_t_v == NULL)
             s_a_t_v = ML_(addStr)(di, "start_according_to_valgrind", -1);
@@ -509,7 +509,7 @@ static Bool file_exists_p(const HChar *path)
 }
 
 
-/* Search for an existing dSYM file as a possible separate debug file.  
+/* Search for an existing dSYM file as a possible separate debug file.
    Adapted from gdb. */
 static HChar *
 find_separate_debug_file (const HChar *executable_name)
@@ -518,7 +518,7 @@ find_separate_debug_file (const HChar *executable_name)
    HChar *dot_ptr;
    HChar *slash_ptr;
    HChar *dsymfile;
-    
+
    /* Make sure the object file name itself doesn't contain ".dSYM" in it or we
       will end up with an infinite loop where after we add a dSYM symbol file,
       it will then enter this function asking if there is a debug file for the
@@ -527,23 +527,23 @@ find_separate_debug_file (const HChar *executable_name)
    {
       /* Check for the existence of a .dSYM file for a given executable.  */
       basename_str = VG_(basename) (executable_name);
-      dsymfile = ML_(dinfo_zalloc)("di.readmacho.dsymfile", 
+      dsymfile = ML_(dinfo_zalloc)("di.readmacho.dsymfile",
                     VG_(strlen) (executable_name)
                     + VG_(strlen) (APPLE_DSYM_EXT_AND_SUBDIRECTORY)
                     + VG_(strlen) (basename_str)
                     + 1
                  );
-        
+
       /* First try for the dSYM in the same directory as the original file.  */
       VG_(strcpy) (dsymfile, executable_name);
       VG_(strcat) (dsymfile, APPLE_DSYM_EXT_AND_SUBDIRECTORY);
       VG_(strcat) (dsymfile, basename_str);
-        
+
       if (file_exists_p (dsymfile))
          return dsymfile;
-        
+
       /* Now search for any parent directory that has a '.' in it so we can find
-         Mac OS X applications, bundles, plugins, and any other kinds of files. 
+         Mac OS X applications, bundles, plugins, and any other kinds of files.
          Mac OS X application bundles wil have their program in
          "/some/path/MyApp.app/Contents/MacOS/MyApp" (or replace ".app" with
          ".bundle" or ".plugin" for other types of bundles).  So we look for any
@@ -568,7 +568,7 @@ find_separate_debug_file (const HChar *executable_name)
             if (file_exists_p (dsymfile))
                return dsymfile;
          }
-         
+
          /* NULL terminate the string at the '.' character and append
             the path down to the dSYM file.  */
          *dot_ptr = '\0';
@@ -576,11 +576,11 @@ find_separate_debug_file (const HChar *executable_name)
          VG_(strcat) (dot_ptr, basename_str);
          if (file_exists_p (dsymfile))
             return dsymfile;
-         
+
          /* NULL terminate the string at the '.' locatated by the strrchr()
             function again.  */
          *dot_ptr = '\0';
-         
+
          /* We found a previous extension '.' character and did not find a
             dSYM file so now find previous directory delimiter so we don't
             try multiple times on a file name that may have a version number
@@ -613,7 +613,7 @@ static DiSlice getsectdata ( DiSlice img,
 
    Int c;
    for (c = 0; c < mh.ncmds; c++) {
-      struct load_command cmd;          
+      struct load_command cmd;
       ML_(cur_read_get)(&cmd, cur, sizeof(cmd));
       if (cmd.cmd == LC_SEGMENT_CMD) {
          struct SEGMENT_COMMAND seg;
@@ -625,7 +625,7 @@ static DiSlice getsectdata ( DiSlice img,
             for (s = 0; s < seg.nsects; s++) {
                struct SECTION sect;
                ML_(cur_step_get)(&sect, &sects_cur, sizeof(sect));
-               if (0 == VG_(strncmp)(sect.sectname, sectname, 
+               if (0 == VG_(strncmp)(sect.sectname, sectname,
                                      sizeof(sect.sectname))) {
                   DiSlice res = img;
                   res.ioff = sect.offset;
@@ -721,7 +721,7 @@ Bool ML_(read_macho_debug_info)( struct _DebugInfo* di )
    const DebugInfoMapping* rx_map = NULL;
    const DebugInfoMapping* rw_map = NULL;
 
-   /* mmap the object file to look for di->soname and di->text_bias 
+   /* mmap the object file to look for di->soname and di->text_bias
       and uuid and nlist */
 
    /* This should be ensured by our caller (that we're in the accept
@@ -774,15 +774,15 @@ Bool ML_(read_macho_debug_info)( struct _DebugInfo* di )
    /* Poke around in the Mach-O header, to find some important
       stuff. */
    // Find LC_SYMTAB and LC_DYSYMTAB, if present.
-   // Read di->soname from LC_ID_DYLIB if present, 
-   //    or from LC_ID_DYLINKER if present, 
+   // Read di->soname from LC_ID_DYLIB if present,
+   //    or from LC_ID_DYLINKER if present,
    //    or use "NONE".
    // Get di->text_bias (aka slide) based on the corresponding LC_SEGMENT
    // Get uuid for later dsym search
 
    di->text_bias = 0;
 
-   { 
+   {
       DiCursor cmd_cur = ML_(cur_from_sli)(msli);
 
       struct MACH_HEADER mh;
@@ -796,13 +796,13 @@ Bool ML_(read_macho_debug_info)( struct _DebugInfo* di )
       for (c = 0; c < mh.ncmds; c++) {
          struct load_command cmd;
          ML_(cur_read_get)(&cmd, cmd_cur, sizeof(cmd));
- 
+
          if (cmd.cmd == LC_SYMTAB) {
             sym_cur = cmd_cur;
-         } 
+         }
          else if (cmd.cmd == LC_DYSYMTAB) {
             dysym_cur = cmd_cur;
-         } 
+         }
          else if (cmd.cmd == LC_ID_DYLIB && mh.filetype == MH_DYLIB) {
             // GrP fixme bundle?
             struct dylib_command dcmd;
@@ -867,7 +867,7 @@ Bool ML_(read_macho_debug_info)( struct _DebugInfo* di )
             if (!di->text_present
                 && 0 == VG_(strcmp)(&seg.segname[0], "__TEXT")
                 /* DDD: is the  next line a kludge? -- JRS */
-                && seg.fileoff == 0 && seg.filesize != 0) {
+                && (from_memory || seg.fileoff == 0) && seg.filesize != 0) {
                di->text_present = True;
                di->text_svma = (Addr)seg.vmaddr;
                di->text_avma = rx_map->avma;
@@ -938,7 +938,7 @@ Bool ML_(read_macho_debug_info)( struct _DebugInfo* di )
 
       ML_(cur_read_get)(&symcmd,   sym_cur,   sizeof(symcmd));
       if (has_dynamic) {
-      ML_(cur_read_get)(&dysymcmd, dysym_cur, sizeof(dysymcmd));
+        ML_(cur_read_get)(&dysymcmd, dysym_cur, sizeof(dysymcmd));
       }
 
       /* Read nlist symbol table */
@@ -957,7 +957,7 @@ Bool ML_(read_macho_debug_info)( struct _DebugInfo* di )
           || msli.szB < symcmd.symoff + symcmd.nsyms * sizeof(struct NLIST))) {
          ML_(symerr)(di, False, "Invalid Mach-O file (5 too small).");
          goto fail;
-      }   
+      }
       if (has_dynamic
           && (dysymcmd.ilocalsym + dysymcmd.nlocalsym > symcmd.nsyms
           || dysymcmd.iextdefsym + dysymcmd.nextdefsym > symcmd.nsyms)) {
@@ -976,10 +976,10 @@ Bool ML_(read_macho_debug_info)( struct _DebugInfo* di )
         strs = ML_(cur_from_sli)(msli);
         strs.ioff = (link_edit_addr + (symcmd.stroff - link_edit.fileoff)) - rx_map->avma;
       } else {
-      syms = ML_(cur_plus)(ML_(cur_from_sli)(msli), symcmd.symoff);
-      strs = ML_(cur_plus)(ML_(cur_from_sli)(msli), symcmd.stroff);
+        syms = ML_(cur_plus)(ML_(cur_from_sli)(msli), symcmd.symoff);
+        strs = ML_(cur_plus)(ML_(cur_from_sli)(msli), symcmd.stroff);
       }
-      
+
       if (VG_(clo_verbosity) > 1) {
         if (has_dynamic) {
          VG_(message)(Vg_DebugMsg,
@@ -1001,18 +1001,18 @@ Bool ML_(read_macho_debug_info)( struct _DebugInfo* di )
                  );
 
       if (has_dynamic) {
-      // extern symbols
-      read_symtab(candSyms,
-                  di,
-                  ML_(cur_plus)(syms,
-                                dysymcmd.iextdefsym * sizeof(struct NLIST)),
-                  dysymcmd.nextdefsym, strs, symcmd.strsize);
-      // static and private_extern symbols
-      read_symtab(candSyms,
-                  di,
-                  ML_(cur_plus)(syms,
-                                dysymcmd.ilocalsym * sizeof(struct NLIST)),
-                  dysymcmd.nlocalsym, strs, symcmd.strsize);
+        // extern symbols
+        read_symtab(candSyms,
+                    di,
+                    ML_(cur_plus)(syms,
+                                  dysymcmd.iextdefsym * sizeof(struct NLIST)),
+                    dysymcmd.nextdefsym, strs, symcmd.strsize);
+        // static and private_extern symbols
+        read_symtab(candSyms,
+                    di,
+                    ML_(cur_plus)(syms,
+                                  dysymcmd.ilocalsym * sizeof(struct NLIST)),
+                    dysymcmd.nlocalsym, strs, symcmd.strsize);
       } else {
         read_symtab(candSyms,
                     di,
@@ -1100,7 +1100,7 @@ Bool ML_(read_macho_debug_info)( struct _DebugInfo* di )
          VG_(message)(Vg_DebugMsg, "%sdSYM directory %s; consider using "
                       "--dsymutil=yes\n",
                       VG_(clo_verbosity) > 1 ? "   " : "",
-                      dsymfilename ? "has wrong UUID" : "is missing"); 
+                      dsymfilename ? "has wrong UUID" : "is missing");
       goto success;
    }
 
@@ -1108,7 +1108,7 @@ Bool ML_(read_macho_debug_info)( struct _DebugInfo* di )
 
    { Int r;
      const HChar* dsymutil = "/usr/bin/dsymutil ";
-     HChar* cmd = ML_(dinfo_zalloc)( "di.readmacho.tmp1", 
+     HChar* cmd = ML_(dinfo_zalloc)( "di.readmacho.tmp1",
                                      VG_(strlen)(dsymutil)
                                      + VG_(strlen)(di->fsm.filename)
                                      + 32 /* misc */ );
@@ -1208,7 +1208,7 @@ Bool ML_(read_macho_debug_info)( struct _DebugInfo* di )
                                          eh_frame_svma + di->text_bias,
                                          True/*is_ehframe*/);
       }
-   
+
       if (ML_(sli_is_valid)(debug_info_mscn)) {
          if (VG_(clo_verbosity) > 1) {
             if (0)

--- a/coregrind/m_debuginfo/readmacho.c
+++ b/coregrind/m_debuginfo/readmacho.c
@@ -37,6 +37,7 @@
 #include "pub_core_libcfile.h"
 #include "pub_core_libcproc.h"
 #include "pub_core_aspacemgr.h"    /* for mmaping debuginfo files */
+#include "pub_core_mach.h"         /* VG_(dyld_cache_get_slide) */
 #include "pub_core_machine.h"      /* VG_ELF_CLASS */
 #include "pub_core_options.h"
 #include "pub_core_oset.h"
@@ -136,7 +137,6 @@ static void unmap_image ( /*MOD*/DiSlice* sli )
    }
 }
 
-
 /* Open the given file, find the thin part if necessary, do some
    checks, and return a DiSlice containing details of both the thin
    part and (implicitly, via the contained DiImage*) the fat part.
@@ -152,7 +152,11 @@ static DiSlice map_image_aboard ( DebugInfo* di, /* only for err msgs */
    DiImage* mimg;
 
    if (rx_map != NULL) {
-     mimg = ML_(img_from_memory)(rx_map->avma, rx_map->size, filename);
+     // FIXME: basically, we make a slice from the place where the mach_header is until the end of the memory space
+     // unfortunately, all the data needed for parsing from the DSC is spread across many places in memory
+     // and there is no way to know for sure the size of the DSC perfectly, so this is the best method at the moment
+     // and it's _very_ unsafe
+     mimg = ML_(img_from_memory)(rx_map->avma, MACH_MEMORY_END - rx_map->avma, filename);
    } else {
      mimg = ML_(img_from_local_file)(filename);
    }
@@ -710,9 +714,10 @@ Bool ML_(read_macho_debug_info)( struct _DebugInfo* di )
    DiCursor dysym_cur    = DiCursor_INVALID;
    HChar*   dsymfilename = NULL;
    Bool     have_uuid    = False;
-   Bool     from_memory  = False;
+   Bool     from_memory  = False; // True if we're reading from DSC
    UChar    uuid[16];
    Word     i;
+   struct SEGMENT_COMMAND     link_edit = {.cmd = 0};
    const DebugInfoMapping* rx_map = NULL;
    const DebugInfoMapping* rw_map = NULL;
 
@@ -744,10 +749,17 @@ Bool ML_(read_macho_debug_info)( struct _DebugInfo* di )
    vg_assert(rx_map);
    vg_assert(rw_map || from_memory);
 
-   if (VG_(clo_verbosity) > 1)
+   if (VG_(clo_verbosity) > 1) {
+      if (from_memory) {
+        VG_(message)(Vg_DebugMsg,
+                    "%s (rx at %#lx)\n", di->fsm.filename,
+                    rx_map->avma);
+      } else {
       VG_(message)(Vg_DebugMsg,
                    "%s (rx at %#lx, rw at %#lx)\n", di->fsm.filename,
                    rx_map->avma, rw_map->avma );
+      }
+   }
 
    VG_(memset)(&uuid, 0, sizeof(uuid));
 
@@ -881,6 +893,10 @@ Bool ML_(read_macho_debug_info)( struct _DebugInfo* di )
                di->data_debug_svma = di->data_svma;
                di->data_debug_bias = di->data_bias;
             }
+            /* Try for __LINKEDIT */
+            if (0 == VG_(strcmp)(&seg.segname[0], "__LINKEDIT")) {
+              link_edit = seg;
+            }
          }
          else if (cmd.cmd == LC_UUID) {
              ML_(cur_read_get)(&uuid, cmd_cur, sizeof(uuid));
@@ -890,6 +906,11 @@ Bool ML_(read_macho_debug_info)( struct _DebugInfo* di )
          cmd_cur = ML_(cur_plus)(cmd_cur, cmd.cmdsize);
       }
    }
+
+  if (from_memory && link_edit.cmd == 0) {
+    ML_(symerr)(di, False, "Invalid Mach-O file (missing __LINKEDIT).");
+    goto fail;
+  }
 
    if (!di->soname) {
       di->soname = ML_(dinfo_strdup)("di.readmacho.noname", "NONE");
@@ -926,9 +947,14 @@ Bool ML_(read_macho_debug_info)( struct _DebugInfo* di )
       XArray* /* DiSym */ candSyms = NULL;
       Word nCandSyms;
 
-      if (msli.szB < symcmd.stroff + symcmd.strsize
-          || msli.szB < symcmd.symoff + symcmd.nsyms
-                                        * sizeof(struct NLIST)) {
+      // FIXME: there is no real nice way to check this when using DSC
+      // the chunk that msli points to is only the mach_header + load_commands
+      // but not the actual sections (__TEXT, __DATA, __LINKEDIT, etc)
+      // so those checks will always fail because they point to data much further in memory
+      // (as DSC groups all this kind of data together for X amount of images)
+      if (!from_memory
+          && (msli.szB < symcmd.stroff + symcmd.strsize
+          || msli.szB < symcmd.symoff + symcmd.nsyms * sizeof(struct NLIST))) {
          ML_(symerr)(di, False, "Invalid Mach-O file (5 too small).");
          goto fail;
       }   
@@ -939,8 +965,20 @@ Bool ML_(read_macho_debug_info)( struct _DebugInfo* di )
          goto fail;
       }
 
+      if (from_memory) {
+        // First, we calculate the real position of __LINKEDIT in the DSC by adding the slide
+        // Then we get the offset of syms/strings within __LINKEDIT by removing the fileoffset
+        // Then we add the __LINKEDIT address
+        // Finally we calculate the proper offset of those addresses compared to the mach_header slice
+        Addr link_edit_addr = link_edit.vmaddr + VG_(dyld_cache_get_slide)();
+        syms = ML_(cur_from_sli)(msli);
+        syms.ioff = (link_edit_addr + (symcmd.symoff - link_edit.fileoff)) - rx_map->avma;
+        strs = ML_(cur_from_sli)(msli);
+        strs.ioff = (link_edit_addr + (symcmd.stroff - link_edit.fileoff)) - rx_map->avma;
+      } else {
       syms = ML_(cur_plus)(ML_(cur_from_sli)(msli), symcmd.symoff);
       strs = ML_(cur_plus)(ML_(cur_from_sli)(msli), symcmd.stroff);
+      }
       
       if (VG_(clo_verbosity) > 1) {
         if (has_dynamic) {

--- a/coregrind/m_libcprint.c
+++ b/coregrind/m_libcprint.c
@@ -326,17 +326,6 @@ void VG_(print_preamble)(Bool logging_to_fd)
       VG_(message)(Vg_DebugMsg,
                    "Valgrind library directory: %s\n", VG_(libdir));
    }
-
-#if defined(VGO_darwin)
-#if DARWIN_VERS >= DARWIN_11_00
-   VG_(message)(Vg_UserMsg,
-    "\n"
-    "Support for macOS 11 and later is currently experimental\n"
-    "Some reports (especially memory leaks) might be missing or incorrect\n"
-    "\n"
-  );
-#endif
-#endif
 }
 
 /* ---------------------------------------------------------------------

--- a/coregrind/m_mach/dyld_cache.c
+++ b/coregrind/m_mach/dyld_cache.c
@@ -32,7 +32,7 @@
 // as some dylib are not provided in file format anymore
 #if defined(VGO_darwin) && DARWIN_VERS >= DARWIN_11_00
 
-#include "pub_core_debuginfo.h"             // VG_(di_notify_mmap)
+#include "pub_core_debuginfo.h"             // VG_(di_notify_dsc)
 #include "pub_core_debuglog.h"              // VG_(debugLog)
 #include "pub_core_mach.h"                  // VG_(dyld_cache_*)
 #include "pub_core_syscall.h"               // VG_(do_syscall1)
@@ -49,6 +49,13 @@
 // FIXME: probably shouldn't include this directly?
 #include "m_syswrap/priv_syswrap-generic.h" // ML_(notify_core_and_tool_of_mmap)
 
+#include <mach-o/loader.h>
+#include <mach-o/fat.h>
+
+// Only supported on macOS 11 onwards which is 64bit only
+# define MACH_HEADER mach_header_64
+# define MAGIC MH_MAGIC_64
+
 static void output_debug_info(const dyld_cache_header* dyld_cache);
 
 typedef struct {
@@ -56,7 +63,10 @@ typedef struct {
   Addr slide;
 } DYLDCache;
 
-static DYLDCache dyld_cache = {};
+static DYLDCache dyld_cache = {
+  .header = NULL,
+  .slide = 0,
+};
 
 static Addr calculate_relative(const dyld_cache_header * header, Addr offset) {
   return (Addr)header + offset;
@@ -200,9 +210,79 @@ void VG_(dyld_cache_init)(void) {
   }
 }
 
-static void output_debug_info(const dyld_cache_header* dyld_cache) {
-  const uint8_t* u1 = dyld_cache->uuid;
-  const uint8_t* u2 = dyld_cache->symbolFileUUID;
+int VG_(dyld_cache_might_be_in)(const HChar* path) {
+  // If not init'd, there is no point
+  if (dyld_cache.header == NULL) {
+    return 0;
+  }
+
+  if (VG_(strncmp)(path, "/usr/lib/", 9) == 0) {
+		return 1;
+  }
+	if (VG_(strncmp)(path, "/System/Library/", 16) == 0) {
+    return 1;
+  }
+  // FIXME: more flexible heuristics around extensions?
+  return 0;
+}
+
+static struct MACH_HEADER* find_image_text(const dyld_cache_header* header, const char* path, SizeT* len) {
+  vg_assert(len);
+  *len = 0;
+
+  const dyld_cache_image_text_info* textInfos = (const dyld_cache_image_text_info*) calculate_relative(header, header->imagesTextOffset);
+
+  for (int i = 0; i < header->imagesTextCount; ++i) {
+    const dyld_cache_image_text_info* textInfo = &textInfos[i];
+    const char* imagePath = (const char*) calculate_relative(header, textInfo->pathOffset);
+
+    if (VG_(strcmp)(imagePath, path) == 0) {
+      *len = textInfo->textSegmentSize;
+      return (struct MACH_HEADER*) calculate_unslid(textInfo->loadAddress);
+    }
+  }
+
+  return NULL;
+}
+
+int VG_(dyld_cache_load_library)(const HChar* path) {
+  struct MACH_HEADER *image = NULL;
+  ULong res = 0;
+  SizeT len = 0;
+
+  // If not init'd, there is no point trying
+  if (dyld_cache.header == NULL) {
+    return 0;
+  }
+
+  VG_(debugLog)(2, "dyld_cache", "potential dylib to check in the cache: %s\n", path);
+
+  image = find_image_text(dyld_cache.header, path, &len);
+  if (image == NULL) {
+    VG_(debugLog)(2, "dyld_cache", "image not found: %s\n", path);
+    return 0;
+  }
+
+  if (image->magic != MAGIC) {
+    VG_(debugLog)(2, "dyld_cache", "image not mach-o (%#x): %s\n", image->magic, path);
+    return 0;
+  }
+
+  VG_(debugLog)(2, "dyld_cache", "image is valid, forwarding to debuginfo: %s\n", path);
+  res = VG_(di_notify_dsc)(path, (Addr)image, &len);
+  if (res == 0) {
+    VG_(debugLog)(2, "dyld_cache", "failed to load debuginfo from: %s\n", path);
+    return 0;
+  }
+
+  VG_(debugLog)(2, "dyld_cache", "image fully loaded: %s\n", path);
+
+  return 1;
+}
+
+static void output_debug_info(const dyld_cache_header* cache) {
+  const uint8_t* u1 = cache->uuid;
+  const uint8_t* u2 = cache->symbolFileUUID;
   VG_(debugLog)(4, "dyld_cache",
     "shared dyld content: {\n"
     "  .magic: %s,\n"
@@ -281,83 +361,83 @@ static void output_debug_info(const dyld_cache_header* dyld_cache) {
     "  .dynamicDataOffset: %#llx,\n"
     "  .dynamicDataMaxSize: %llu,\n"
     "}\n",
-    dyld_cache->magic,
-    dyld_cache->mappingOffset,
-    dyld_cache->mappingCount,
-    dyld_cache->imagesOffsetOld,
-    dyld_cache->imagesCountOld,
-    dyld_cache->dyldBaseAddress,
-    dyld_cache->codeSignatureOffset,
-    dyld_cache->codeSignatureSize,
-    dyld_cache->slideInfoOffsetUnused,
-    dyld_cache->slideInfoSizeUnused,
-    dyld_cache->localSymbolsOffset,
-    dyld_cache->localSymbolsSize,
+    cache->magic,
+    cache->mappingOffset,
+    cache->mappingCount,
+    cache->imagesOffsetOld,
+    cache->imagesCountOld,
+    cache->dyldBaseAddress,
+    cache->codeSignatureOffset,
+    cache->codeSignatureSize,
+    cache->slideInfoOffsetUnused,
+    cache->slideInfoSizeUnused,
+    cache->localSymbolsOffset,
+    cache->localSymbolsSize,
     u1[0], u1[1], u1[2], u1[3], u1[4], u1[5], u1[6], u1[7], u1[8],
     u1[9], u1[10], u1[11], u1[12], u1[13], u1[14], u1[15],
-    dyld_cache->cacheType,
-    dyld_cache->branchPoolsOffset,
-    dyld_cache->branchPoolsCount,
-    dyld_cache->dyldInCacheMH,
-    dyld_cache->dyldInCacheEntry,
-    dyld_cache->imagesTextOffset,
-    dyld_cache->imagesTextCount,
-    dyld_cache->patchInfoAddr,
-    dyld_cache->patchInfoSize,
-    dyld_cache->otherImageGroupAddrUnused,
-    dyld_cache->otherImageGroupSizeUnused,
-    dyld_cache->progClosuresAddr,
-    dyld_cache->progClosuresSize,
-    dyld_cache->progClosuresTrieAddr,
-    dyld_cache->progClosuresTrieSize,
-    dyld_cache->platform,
-    dyld_cache->formatVersion,
-    dyld_cache->dylibsExpectedOnDisk,
-    dyld_cache->simulator,
-    dyld_cache->locallyBuiltCache,
-    dyld_cache->builtFromChainedFixups,
-    dyld_cache->padding,
-    dyld_cache->sharedRegionStart,
-    dyld_cache->sharedRegionSize,
-    dyld_cache->maxSlide,
-    dyld_cache->dylibsImageArrayAddr,
-    dyld_cache->dylibsImageArraySize,
-    dyld_cache->dylibsTrieAddr,
-    dyld_cache->dylibsTrieSize,
-    dyld_cache->otherImageArrayAddr,
-    dyld_cache->otherImageArraySize,
-    dyld_cache->otherTrieAddr,
-    dyld_cache->otherTrieSize,
-    dyld_cache->mappingWithSlideOffset,
-    dyld_cache->mappingWithSlideCount,
-    dyld_cache->dylibsPBLStateArrayAddrUnused,
-    dyld_cache->dylibsPBLSetAddr,
-    dyld_cache->programsPBLSetPoolAddr,
-    dyld_cache->programsPBLSetPoolSize,
-    dyld_cache->programTrieAddr,
-    dyld_cache->programTrieSize,
-    dyld_cache->osVersion,
-    dyld_cache->altPlatform,
-    dyld_cache->altOsVersion,
-    dyld_cache->swiftOptsOffset,
-    dyld_cache->swiftOptsSize,
-    dyld_cache->subCacheArrayOffset,
-    dyld_cache->subCacheArrayCount,
+    cache->cacheType,
+    cache->branchPoolsOffset,
+    cache->branchPoolsCount,
+    cache->dyldInCacheMH,
+    cache->dyldInCacheEntry,
+    cache->imagesTextOffset,
+    cache->imagesTextCount,
+    cache->patchInfoAddr,
+    cache->patchInfoSize,
+    cache->otherImageGroupAddrUnused,
+    cache->otherImageGroupSizeUnused,
+    cache->progClosuresAddr,
+    cache->progClosuresSize,
+    cache->progClosuresTrieAddr,
+    cache->progClosuresTrieSize,
+    cache->platform,
+    cache->formatVersion,
+    cache->dylibsExpectedOnDisk,
+    cache->simulator,
+    cache->locallyBuiltCache,
+    cache->builtFromChainedFixups,
+    cache->padding,
+    cache->sharedRegionStart,
+    cache->sharedRegionSize,
+    cache->maxSlide,
+    cache->dylibsImageArrayAddr,
+    cache->dylibsImageArraySize,
+    cache->dylibsTrieAddr,
+    cache->dylibsTrieSize,
+    cache->otherImageArrayAddr,
+    cache->otherImageArraySize,
+    cache->otherTrieAddr,
+    cache->otherTrieSize,
+    cache->mappingWithSlideOffset,
+    cache->mappingWithSlideCount,
+    cache->dylibsPBLStateArrayAddrUnused,
+    cache->dylibsPBLSetAddr,
+    cache->programsPBLSetPoolAddr,
+    cache->programsPBLSetPoolSize,
+    cache->programTrieAddr,
+    cache->programTrieSize,
+    cache->osVersion,
+    cache->altPlatform,
+    cache->altOsVersion,
+    cache->swiftOptsOffset,
+    cache->swiftOptsSize,
+    cache->subCacheArrayOffset,
+    cache->subCacheArrayCount,
     u2[0], u2[1], u2[2], u2[3], u2[4], u2[5], u2[6], u2[7], u2[8],
     u2[9], u2[10], u2[11], u2[12], u2[13], u2[14], u2[15],
-    dyld_cache->rosettaReadOnlyAddr,
-    dyld_cache->rosettaReadOnlySize,
-    dyld_cache->rosettaReadWriteAddr,
-    dyld_cache->rosettaReadWriteSize,
-    dyld_cache->imagesOffset,
-    dyld_cache->imagesCount,
-    dyld_cache->cacheSubType,
-    dyld_cache->objcOptsOffset,
-    dyld_cache->objcOptsSize,
-    dyld_cache->cacheAtlasOffset,
-    dyld_cache->cacheAtlasSize,
-    dyld_cache->dynamicDataOffset,
-    dyld_cache->dynamicDataMaxSize
+    cache->rosettaReadOnlyAddr,
+    cache->rosettaReadOnlySize,
+    cache->rosettaReadWriteAddr,
+    cache->rosettaReadWriteSize,
+    cache->imagesOffset,
+    cache->imagesCount,
+    cache->cacheSubType,
+    cache->objcOptsOffset,
+    cache->objcOptsSize,
+    cache->cacheAtlasOffset,
+    cache->cacheAtlasSize,
+    cache->dynamicDataOffset,
+    cache->dynamicDataMaxSize
   );
 }
 

--- a/coregrind/m_mach/dyld_cache.c
+++ b/coregrind/m_mach/dyld_cache.c
@@ -200,6 +200,10 @@ static int try_to_init(void) {
   return 1;
 }
 
+Addr VG_(dyld_cache_get_slide)(void) {
+  return dyld_cache.slide;
+}
+
 void VG_(dyld_cache_init)(void) {
   if (!try_to_init()) {
     VG_(dmsg)(
@@ -269,7 +273,7 @@ int VG_(dyld_cache_load_library)(const HChar* path) {
   }
 
   VG_(debugLog)(2, "dyld_cache", "image is valid, forwarding to debuginfo: %s\n", path);
-  res = VG_(di_notify_dsc)(path, (Addr)image, &len);
+  res = VG_(di_notify_dsc)(path, (Addr)image, len);
   if (res == 0) {
     VG_(debugLog)(2, "dyld_cache", "failed to load debuginfo from: %s\n", path);
     return 0;

--- a/coregrind/m_mach/dyld_cache.c
+++ b/coregrind/m_mach/dyld_cache.c
@@ -221,9 +221,9 @@ int VG_(dyld_cache_might_be_in)(const HChar* path) {
   }
 
   if (VG_(strncmp)(path, "/usr/lib/", 9) == 0) {
-		return 1;
+    return 1;
   }
-	if (VG_(strncmp)(path, "/System/Library/", 16) == 0) {
+  if (VG_(strncmp)(path, "/System/Library/", 16) == 0) {
     return 1;
   }
   // FIXME: more flexible heuristics around extensions?

--- a/coregrind/m_mach/priv_dyld_internals.h
+++ b/coregrind/m_mach/priv_dyld_internals.h
@@ -122,4 +122,12 @@ typedef struct {
   uint64_t    cacheVMOffset;      // The offset of this subcache from the main cache base address
 } dyld_subcache_entry_v1;
 
+// From Apple's `dyld/cache-builder/dyld_cache_format.h`
+typedef struct {
+  uuid_t      uuid;
+  uint64_t    loadAddress;            // unslid address of start of __TEXT
+  uint32_t    textSegmentSize;
+  uint32_t    pathOffset;             // offset from start of cache file
+} dyld_cache_image_text_info;
+
 #endif

--- a/coregrind/m_syswrap/syswrap-darwin.c
+++ b/coregrind/m_syswrap/syswrap-darwin.c
@@ -44,6 +44,7 @@
 #include "pub_core_libcprint.h"
 #include "pub_core_libcproc.h"
 #include "pub_core_libcsignal.h"
+#include "pub_core_mach.h"         // VG_(dyld_cache_*)
 #include "pub_core_machine.h"      // VG_(get_SP)
 #include "pub_core_mallocfree.h"
 #include "pub_core_options.h"
@@ -3080,10 +3081,34 @@ PRE(stat64)
    PRE_REG_READ2(long, "stat", const char *,path, struct stat64 *,buf);
    PRE_MEM_RASCIIZ("stat64(path)", ARG1);
    PRE_MEM_WRITE( "stat64(buf)", ARG2, sizeof(struct vki_stat64) );
+
+#if DARWIN_VERS >= DARWIN_11_00
+   // Starting with macOS 11.0, some system libraries are not provided on the disk but only though
+   // shared dyld cache, thus we try to detect if dyld tried (and failed) to load a dylib,
+   // in which case we do the same thing as dyld and load the info from the cache directly
+   //
+   // This is our entry point for checking a particular dylib: if it looks like one,
+   // we want to see the error result, if any, and subsequently check the cache
+   if (VG_(dyld_cache_might_be_in)((HChar *)ARG1)) {
+     *flags |= SfPostOnFail;
+   }
+#endif
 }
 POST(stat64)
 {
+   if (SUCCESS) {
    POST_MEM_WRITE( ARG2, sizeof(struct vki_stat64) );
+}
+
+#if DARWIN_VERS >= DARWIN_11_00
+   if (SUCCESS || (FAILURE && ERR == VKI_ENOENT)) {
+     // It failed and `SfPostOnFail` was set, thus this is probably a dylib,
+     // try to load it from cache which will call VG_(di_notify_mmap) like the previous versions did
+     if (VG_(dyld_cache_load_library)((HChar *)ARG1)) {
+       ML_(sync_mappings)("after", "stat64", 0);
+     }
+   }
+#endif
 }
 
 PRE(lstat64)

--- a/coregrind/pub_core_debuginfo.h
+++ b/coregrind/pub_core_debuginfo.h
@@ -76,6 +76,10 @@ extern void VG_(di_notify_pdb_debuginfo)( Int fd, Addr avma,
 extern void VG_(di_notify_vm_protect)( Addr a, SizeT len, UInt prot );
 #endif
 
+#if defined(VGO_darwin) && DARWIN_VERS >= DARWIN_11_00
+extern ULong VG_(di_notify_dsc)( const HChar* path, Addr header, SizeT len );
+#endif
+
 extern void VG_(di_discard_ALL_debuginfo)( void );
 
 /* Like VG_(get_fnname), but it does not do C++ demangling nor Z-demangling

--- a/coregrind/pub_core_mach.h
+++ b/coregrind/pub_core_mach.h
@@ -39,12 +39,21 @@
 // Call this early in Valgrind's main(). It depends on nothing.
 extern void VG_(mach_init)(void);
 
+#if defined(VGP_amd64_darwin)
+#define MACH_MEMORY_END 0x7ffffffff000
+#elif defined(VGP_x86_darwin)
+#define MACH_MEMORY_END 0xfffff000
+#else
+#error "Unsupported platform"
+#endif
+
 #if DARWIN_VERS >= DARWIN_11_00
 // Dyld shared cache (DSC) parsing, which is required as system libraries are not provided on disk
 // starting with macOS 11.0 (Big Sur)
 extern void VG_(dyld_cache_init)(void);
 extern int VG_(dyld_cache_might_be_in)(const HChar*);
 extern int VG_(dyld_cache_load_library)(const HChar*);
+extern Addr VG_(dyld_cache_get_slide)(void);
 #endif
 
 #endif // __PUB_CORE_MACH_H

--- a/coregrind/pub_core_mach.h
+++ b/coregrind/pub_core_mach.h
@@ -40,10 +40,11 @@
 extern void VG_(mach_init)(void);
 
 #if DARWIN_VERS >= DARWIN_11_00
-// Component replicating dyld shared cache checking
-// as system libraries are not provided on disk
+// Dyld shared cache (DSC) parsing, which is required as system libraries are not provided on disk
 // starting with macOS 11.0 (Big Sur)
 extern void VG_(dyld_cache_init)(void);
+extern int VG_(dyld_cache_might_be_in)(const HChar*);
+extern int VG_(dyld_cache_load_library)(const HChar*);
 #endif
 
 #endif // __PUB_CORE_MACH_H

